drivers/flash/spi_nor.c

```diff
@@ -568,10 +577,25 @@ static const struct flash_driver_api spi_nor_api = {
 
 static const struct spi_nor_config flash_id = {
 	.id = DT_INST_0_JEDEC_SPI_NOR_JEDEC_ID,
-#ifdef DT_INST_0_JEDEC_SPI_NOR_HAS_BE32K
-	.has_be32k = true,
-#endif /* DT_INST_0_JEDEC_SPI_NOR_HAS_BE32K */
 	.size = DT_INST_0_JEDEC_SPI_NOR_SIZE / 8,
+
+/* defaults for page size and erase types */
+	.page_size = DT_INST_0_JEDEC_SPI_NOR_PAGE_SIZE,
+	.erase_types = { { DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_0,
+			   DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_1 },
+			 { DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_2,
+			   DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_3 },
+			 { DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_4,
+			   DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_5 },
+			 { DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_6,
+			   DT_INST_0_JEDEC_SPI_NOR_ERASE_TYPES_7 } },
+
+/* device specific values */
+	COND_CODE_1(DT_INST_0_CYPRESS_S25FL512S, (
```

Comment:
I'm not convinced this (hard-coding a set of configurations based on compatible) is the right approach, at least not until we can identify supported compatibles independent of instance number and have code to iterate through configurations looking up a JEDEC ID.  If some device has four of the same SPI chip we shouldn't need to provide its configuration four times.

Devicetree support needed (`#if DT_HAS_COMPAT(cypress_s25fl512s)`) may be coming later in this release via #23245, but even that isn't quite right.  We really want some way to do the runtime lookup based on a table that is *not* derived from devicetree (e.g. for firmware that works on multiple versions of a board that differ only in the installed flash chip).  We're not there yet, so for now I think we should just have the devicetree node specify the parameters for each instance, allowing the default to fill in when the node doesn't provide overrides.

Reply:
> I'm not convinced this (hard-coding a set of configurations based on compatible) is the right approach, at least not until we can identify supported compatibles independent of instance number and have code to iterate through configurations looking up a JEDEC ID. If some device has four of the same SPI chip we shouldn't need to provide its configuration four times.
> 
> Devicetree support needed (`#if DT_HAS_COMPAT(cypress_s25fl512s)`) may be coming later in this release via #23245, but even that isn't quite right. We really want some way to do the runtime lookup based on a table that is _not_ derived from devicetree (e.g. for firmware that works on multiple versions of a board that differ only in the installed flash chip).

If this is the goal (same firmware, different flash devices with different configuration) wouldn't it be easier to just query the SFDP during runtime? Otherwise we need to provide the (maybe quite big) configuration for all supported, non default chips in flash, bloating the image size. It certainly would have to be an optional feature and we would need some way to specify which chips configuration should be compiled in.

> We're not there yet, so for now I think we should just have the devicetree node specify the parameters for each instance, allowing the default to fill in when the node doesn't provide overrides.

I removed the conditional code as you suggested.

Reply:
> If this is the goal (same firmware, different flash devices with different configuration) wouldn't it be easier to just query the SFDP during runtime?

SFDP is what you want for firmware that has to run on multiple platforms where you can't anticipate what flash chips may be present.  The alternative for that use case is a big table of all known devices, which I agree isn't viable.

The use case I'm considering is a product line where variant 1 goes out with a 16 MiBi flash, then variant 2 goes out with a 128 MiBi flash.  The firmware supporting v2 should run on v1.  That can be done by having it look at the JEDEC id and redirect a pointer to the interface constants for the device that's present.

What's needed then is some way to provide the table of potential devices to the driver.  One way would be a weak definition something like:

```
__weak struct spi_nor_config *flash_spi_nor_config_for_device(struct device *dev,
                                                              const u8_t *jedec_id)
{
    return dev->config->config_info;
}
```
and modify `spi_nor_read_id()` to store in `*data` a pointer to the params that override the default in `*config`.  All under a Kconfig enabling the feature.

Then the application need merely provide a custom implementation that works for the flash devices it supports.  No size impact on applications that don't need it; minimal impact on those that do.

But we don't need to implement that feature right now.

Reply:
Thank you for clarification.